### PR TITLE
Create edit_reports permission for configurable reports

### PR DESCRIPTION
**Pull request by solleks**
This is a copy of an [earlier draft PR](https://github.com/dimagi/commcare-hq/pull/27993) that I messed up due to my ineptness with git.

[Feature request document.](https://docs.google.com/document/d/1bmo7swOiXyUTW6IcWn9HFCNcdUS8m6KzMcSK2biN7zg/edit?ts=5ef3962a)

[Trello card.](https://trello.com/c/YjJacM4G/83-detach-reports-from-data-permissions)

##### SUMMARY
This change creates a new permission ‘edit_reports’ that controls two actions - creating custom reports and editing custom reports. The pieces of the change are:

- Add the ‘edit_reports’ property to the corehq.apps.users.models.Permissions class.
- Add a migration for the new property that copies the ‘edit_data’ property.
- Change the following permissions checks from ‘edit_data’ to ‘edit_reports’
    - Whether the “Report Builder” section appears in the ProjectReportsTab
    - The method_decorator on ReportBuilderView.dispatch
- Propagate ‘edit_reports’ to page_context of ConfigurableReportView to avoid showing the ‘Edit Report’ button.
- Add ‘Reports’ section to edit role modal dialog, which contains Create and Edit Reports checkbox for ‘edit_reports’ permission and existing Access All Reports and Access Specific Reports controls.
EditRole
- Change Reports column of project ‘settings/users/roles/’ page to show a check mark when role has ‘edit_reports’ permission and a view only icon when role has only ‘view_reports’ permission.
RolePermissions

##### FEATURE FLAG
Not specific to a feature flag.

##### RISK ASSESSMENT / QA PLAN
This change should get a QA plan as it will affect users permissions to take actions.
I need to learn how to set up a QA plan.

##### PRODUCT DESCRIPTION
This change allows project administrators to create roles that can edit data but not create/edit reports and vice versa. They will make the choice in the edit role dialog and see the effect in the role and permissions page.

### CODE REVIEW

**Comment by solleks** [comment]
proteusvacuum reviewed my earlier draft PR of this change. I will copy his comments into this version.

**Comment by solleks** [comment]
From proteusvacuum:

Did you mean to commit the submodule changes? Sometimes these diverge from the branch you are working on, and it doesn't seem like the changes in those are related to this PR.

To make a QA request you can fill out this form: https://docs.google.com/forms/d/e/1FAIpQLSehQl5VXuHED_HLXaMcUoWleBGqpZu_chPXfWaxdcb0QbDh4Q/formResponse and the QA team will reach out to you if they need clarifications.

We normally deploy QA branches to staging - instructions for how to do that are found in scripts/staging.yaml, but feel free to reach out in #staging on slack if you have questions about that, I'm not sure if you have deploy permissions on the staging machines.

Feel free to reach out to me on slack if you wanted to chat through anything else, although this seems pretty complete already.

**Comment by solleks** [comment]
The submodule changes were not intentional, so I removed them.

I don't have permission to open the QA request form. Cal suggested that I contact Ben Talbot, so I will do that separately.

I'm sure that I don't have permissions to touch anything in staging.

File: corehq/apps/users/migrations/0019_editreportspermissions.py
Comment (solleks) [inline corehq/apps/users/migrations/0019_editreportspermissions.py:23]: proteusvacuum commented "I'm not sure I follow why this is being set to False here."

I borrowed this logic from the web apps permission change, which was added after the initial change:
https://github.com/dimagi/commcare-hq/blame/master/corehq/apps/users/migrations/0016_webappspermissions.py

It makes sure that edit_reports won't be true if edit_data is false.

File: corehq/apps/users/models.py
Comment (solleks) [inline corehq/apps/users/models.py:2713]: I added this so that a unit test would work, but I want to call attention to this change to make sure it's alright.

Comment (esoergel) [reply corehq/apps/users/models.py:2713]: This should be fine, though I don't think the tests should be using `AnonymousCouchUser`s (more about that above)

File: corehq/apps/userreports/tests/test_view.py
Comment (solleks) [reply corehq/apps/userreports/tests/test_view.py]: Thanks for the pointers. I've changed the test to use a WebUser instead of the AnonymousCouchUser.

File: corehq/apps/userreports/tests/test_view.py
Comment (esoergel) [inline corehq/apps/userreports/tests/test_view.py:220]: Can you also make sure the test cleans up these objects?  `self.addCleanup(domain.delete)` is probably the easiest way to do it.  That should also cascade delete the roles, but the web users will need to be deleted individually.

File: corehq/apps/userreports/tests/test_view.py
Comment (esoergel) [inline corehq/apps/userreports/tests/test_view.py:251]:
```suggestion
            request.user = web_user.get_django_user()
```

File: corehq/apps/userreports/tests/test_view.py
Comment (solleks) [reply corehq/apps/userreports/tests/test_view.py:220]: Sure, I've added those calls to addCleanup.

File: corehq/apps/userreports/tests/test_view.py
Comment (solleks) [reply corehq/apps/userreports/tests/test_view.py]: Done.

**Comment by solleks** [comment]
Could you add the label "product/all-users-all-environments" to this request? I don't seem to be able to add a label.

File: corehq/apps/users/templates/users/roles_and_permissions.html
Comment (calellowitz) [inline corehq/apps/users/templates/users/roles_and_permissions.html:167]: It may not be necessary to add this to the default page. Specifically, we appear to generally exclude permissions that require a specific subscription level, like this one, since they are confusing for projects that do not have it. Other new permissions like the web apps one are likewise excluded from this page.

File: corehq/apps/users/templates/users/partials/edit_role_modal.html
Comment (calellowitz) [inline corehq/apps/users/templates/users/partials/edit_role_modal.html:519]: I think this should be hidden if the user's subscription does not have access to the report builder feature

**Review by calellowitz — APPROVED** [review]
Generally this looks great (and the screenshots look good as well). Few small comments but thanks for doing this

File: corehq/apps/users/templates/users/roles_and_permissions.html
Comment (solleks) [reply corehq/apps/users/templates/users/roles_and_permissions.html:167]: Let me make sure I understand. I don't think you're saying we should remove the Reports column from this table because it was there before this change. Rather, you're saying that I should just undo this change so the column shows either a check mark, a list of specific reports or empty, but doesn't distinguish the ability to edit from the ability to view. Does that sound right?

File: corehq/apps/users/templates/users/partials/edit_role_modal.html
Comment (solleks) [reply corehq/apps/users/templates/users/partials/edit_role_modal.html:519]: How can we implement that check?

File: corehq/apps/users/templates/users/roles_and_permissions.html
Comment (calellowitz) [reply corehq/apps/users/templates/users/roles_and_permissions.html:167]: Ah, that is an excellent catch, and I definitely missed that it was there before. After this change is it possible for a user to be able to be able to create reports, but not view all reports? What is shown if they can only see some of the reports? It looks like the code that lists reports is still there, do they see the eye and the list? That might be crowded. Depending on the answer to the first question, I think the original behavior is fine, or if its not possible to edit but be restricted, having icons for edit and view all, and the list for restricted seems fine as well.

File: corehq/apps/users/templates/users/partials/edit_role_modal.html
Comment (calellowitz) [reply corehq/apps/users/templates/users/partials/edit_role_modal.html:519]: It looks like this util function does so https://github.com/dimagi/commcare-hq/blob/a37dc36631dbff5f5c63a9d9b49b83a16ded1221/corehq/apps/userreports/util.py#L49

File: corehq/apps/userreports/views.py
Comment (solleks) [reply corehq/apps/userreports/views.py:312]: OK, I've added the checks that make sense to me:
    o login and domain required
    o UCR toggle set to on
    o has edit_reports permission

File: corehq/apps/users/templates/users/partials/edit_role_modal.html
Comment (solleks) [reply corehq/apps/users/templates/users/partials/edit_role_modal.html]: I've added "in report builder" to the label in the form-check (i.e. on the right) but left it out of the label on the form-group because it wrapped and looked out of place with the other labels in its column.

File: corehq/apps/users/templates/users/roles_and_permissions.html
Comment (solleks) [reply corehq/apps/users/templates/users/roles_and_permissions.html]: It is possible for a user to be able to create reports but not view all reports. In that case, they would see both a check mark and a list of reports. Removing the ability to create reports makes it show just the list, with no eye icon (not really sure why). Giving access to all reports without the ability to create them does show the eye icon with "view only".

I propose that we restore this to the way it was - a check mark when users can access all reports and a list when they can only access specific ones. Since the builtin reports are available with any subscription, this is consistent with your first comment. The ability to access reports is available to all so it makes sense in this table, while the ability to create reports requires pro or higher subscriptions so we won't try to show it in this table.

File: corehq/apps/users/templates/users/partials/edit_role_modal.html
Comment (solleks) [reply corehq/apps/users/templates/users/partials/edit_role_modal.html]: Thanks for the pointer. I've added has_report_builder_access to the context and conditionalized this item in the dialog.

**Comment by solleks** [comment]
I added one small change to hide the Create and Edit Reports setting when Full Organization Access is deselected. Location-restricted users aren't able to create custom reports now with 'edit_data' and won't be able to with 'edit_reports' either, and hiding the setting makes this more apparent.

**Comment by solleks** [comment]
I made some changes after Cal's approval and I'd be happy to get feedback on those if you have any.

One QA pass has finished successfully, but Ben said he would be doing another pass. Once that is done and assuming no issues arise with my post-approval changes, I will ask one of you reviewers to perform the merge.

The [documentation](https://confluence.dimagi.com/display/commcarepublic/Roles+and+Permissions) is updated in draft form if anyone wants to take look. That includes adding a description of the Web Apps permission.

**Comment by solleks** [comment]
The QA review has completed with no further issues (https://dimagi-dev.atlassian.net/browse/QA-1515).

Cal or Ethan, if my most recent changes look good to you, could you perform the merge? I do not have permission to do it.

File: corehq/apps/users/templates/users/partials/edit_role_modal.html
Comment (calellowitz) [inline corehq/apps/users/templates/users/partials/edit_role_modal.html:519]: If a user removes full location access, it looks like this permission will still remain. I think if we need to restrict this only to roles with full access, then we will need to make sure to reset this permission when the other is removed

**Review by calellowitz — CHANGES_REQUESTED** [review]
Thanks for making those changes. One thought about the last commit, but otherwise looks good

File: corehq/apps/users/templates/users/partials/edit_role_modal.html
Comment (solleks) [reply corehq/apps/users/templates/users/partials/edit_role_modal.html:519]: I agree that it should probably behave the way you suggest. But it also looks like the existing permissions that depend on access_all_locations (e.g. Web Users and Groups) are not reset either.

I don't really know Knockout.js, so I'd be reluctant to attempt this in the dialog box itself. It looks like it could be done in the 'post_user_role' function though.

Comment (solleks) [reply corehq/apps/users/templates/users/partials/edit_role_modal.html:519]: I added code to disable permissions for location-restricted users. There was a test failure due to new arguments in a method, so I've synced my branch and added a fix.

**Comment by solleks** [comment]
I added permissions checks to the undelete_report view to match delete_view. I also removed the decorator to require UCR because the function _assert_report_delete_privileges is already used.

**Review by calellowitz — APPROVED** [review]
Looks good! Thanks for the followup

**Comment by solleks** [comment]
Thanks. And I will need you or someone else with write access to perform the merge for me.

**Pull request merged**
